### PR TITLE
OGM-983. Changes from Epam

### PR DESCRIPTION
**Pull request by schernolyas**
The PR contains 'automatic schema generation' functionality

### CODE REVIEW

**Comment by schernolyas** [comment]
Hi @DavideD !

Could you please review the PR?

File: ignite/src/main/java/org/hibernate/ogm/datastore/ignite/impl/IgniteCacheInitializer.java [hunk 1]
```diff
@@ -138,40 +183,87 @@ private CacheConfiguration createCacheConfiguration(AssociationKeyMetadata assoc
...
+		if ( tableOptional.isPresent() ) {
+			Table table = tableOptional.get();
+			for ( Iterator<Column> columnIterator = table.getColumnIterator(); columnIterator.hasNext(); ) {
+				Column currentColumn = columnIterator.next();
+				Type valueType = currentColumn.getValue().getType();
+				Value value = currentColumn.getValue();
+				if ( valueType instanceof CustomType ) {
+					CustomType type = (CustomType) currentColumn.getValue().getType();
+					UserType userType = type.getUserType();
+					if ( userType instanceof EnumType ) {
```
Comment (DavideD) [inline ignite/src/main/java/org/hibernate/ogm/datastore/ignite/impl/IgniteCacheInitializer.java:245]: Shouldn't this have an else caluse in case the userType is not an EnumType? Can that Happen?

File: ignite/src/main/java/org/hibernate/ogm/datastore/ignite/impl/IgniteCacheInitializer.java [hunk 2]
```diff
@@ -138,40 +183,87 @@ private CacheConfiguration createCacheConfiguration(AssociationKeyMetadata assoc
...
+				}
+				else if ( valueType instanceof YesNoType ) {
+					queryEntity.addQueryField( currentColumn.getName(), String.class.getName(), null );
+				}
+				else if ( value.getClass() == SimpleValue.class ) {
+					// it is simple type. add the field
+					SimpleValue simpleValue = (SimpleValue) value;
+					Class returnValue = simpleValue.getType().getReturnedClass();
+					returnValue = h2TypeMapping.getOrDefault( returnValue, returnValue );
+					queryEntity.addQueryField( currentColumn.getName(), returnValue.getName(), null );
```
Comment (DavideD) [inline ignite/src/main/java/org/hibernate/ogm/datastore/ignite/impl/IgniteCacheInitializer.java:263]: I thinnk all this part can be removed by using something like:

                TypeTranslator translator = serviceRegistry.getService( TypeTranslator.class );
		Type valueType = currentColumn.getValue().getType();
		GridType gridType = translator.getType( valueType )

But I'm a bit tired, I will continue tomorrow

File: ignite/src/main/java/org/hibernate/ogm/datastore/ignite/impl/IgniteCacheInitializer.java (hunk 1 above)
Comment (schernolyas) [reply ignite/src/main/java/org/hibernate/ogm/datastore/ignite/impl/IgniteCacheInitializer.java:245]: Hi! @DavideD !
I don't know. I got the code from my OrientDB implementation. Do you offer something else ?

Comment (DavideD) [reply ignite/src/main/java/org/hibernate/ogm/datastore/ignite/impl/IgniteCacheInitializer.java:245]: See my other comment

File: ignite/src/main/java/org/hibernate/ogm/datastore/ignite/impl/IgniteCacheInitializer.java (hunk 2 above)
Comment (DavideD) [reply ignite/src/main/java/org/hibernate/ogm/datastore/ignite/impl/IgniteCacheInitializer.java:263]: This solution doesn't seem to work, I'll play a bit more. My concern is that the logic here is  not taking into account how the dialect can change the way values are stored

File: ignite/src/main/java/org/hibernate/ogm/datastore/ignite/impl/IgniteCacheInitializer.java [hunk 3]
```diff
@@ -138,40 +183,87 @@ private CacheConfiguration createCacheConfiguration(AssociationKeyMetadata assoc
 
 	private void appendIndex(QueryEntity queryEntity, AssociationKeyMetadata associationKeyMetadata, SchemaDefinitionContext context) {
 
-		if ( associationKeyMetadata.getColumnNames().length > 1 ) {
-			//composite id. not yet implemented
-			return;
+		for ( String idFieldName : associationKeyMetadata.getRowKeyColumnNames() ) {
+			queryEntity.addQueryField( generateIndexName( idFieldName ), String.class.getName(),null );
```
Comment (DavideD) [inline ignite/src/main/java/org/hibernate/ogm/datastore/ignite/impl/IgniteCacheInitializer.java:187]: Can you explain me what's going on here? How come this is always a String field?

File: ignite/src/test/resources/log4j.properties [hunk 4]
```diff
@@ -8,8 +8,8 @@
 log4j.appender.stdout=org.apache.log4j.ConsoleAppender
 log4j.appender.stdout.Target=System.out
 log4j.appender.stdout.layout=org.apache.log4j.PatternLayout
-log4j.appender.stdout.layout.ConversionPattern=%d{ABSOLUTE} %5p %c{1}:%L - %m%n
+log4j.appender.stdout.layout.ConversionPattern=%d{ABSOLUTE} %5p %c{1}:%M:%L - %m%n
 
 log4j.rootLogger=info, stdout
 
-log4j.logger.org.hibernate=info
+log4j.logger.org.hibernate=debug
```
Comment (DavideD) [inline ignite/src/test/resources/log4j.properties]: This changes should be in a separate commit, you are only changing the properties of the log file

File: ignite/src/main/java/org/hibernate/ogm/datastore/ignite/impl/IgniteCacheInitializer.java (hunk 3 above)
Comment (schernolyas) [reply ignite/src/main/java/org/hibernate/ogm/datastore/ignite/impl/IgniteCacheInitializer.java:187]: I generate fields for support association if it needs.

File: ignite/src/test/resources/log4j.properties (hunk 4 above)
Comment (schernolyas) [reply ignite/src/test/resources/log4j.properties]: You right.  I will restore it. It needs for development and testing

Comment (schernolyas) [reply ignite/src/test/resources/log4j.properties]: fixed

File: ignite/src/main/java/org/hibernate/ogm/datastore/ignite/impl/IgniteCacheInitializer.java (hunk 3 above)
Comment (schernolyas) [reply ignite/src/main/java/org/hibernate/ogm/datastore/ignite/impl/IgniteCacheInitializer.java:187]: It is part of method "appendIndex". It uses for create indexes for joins.

Comment (DavideD) [reply ignite/src/main/java/org/hibernate/ogm/datastore/ignite/impl/IgniteCacheInitializer.java:187]: I can see that but why does it pass String.class.getName as field type, couldn't the associaton field have some other type?

**Comment by DavideD** [comment]
Merged https://github.com/hibernate/hibernate-ogm-ignite/pull/7

I've merged this with an additional commit that make the code easier to understand and keep track of the changes the dialect can make: https://github.com/hibernate/hibernate-ogm-ignite/commit/45a6dbc53a5cc8ad525d715b825ff04c3ef90759

I've also squashed together a few commits.

@schernolyas le me know if something is not right.
